wasmbinding/stargate_whitelist.go

```diff
@@ -33,14 +33,15 @@ import (
 	epochtypes "github.com/osmosis-labs/osmosis/x/epochs/types"
 )
 
-// stargateWhitelist keeps whitelist and its deterministic
+// stargateResponsePools keeps whitelist and its deterministic
 // response binding for stargate queries.
 // CONTRACT: since results of queries go into blocks, queries being added here should always be
 // deterministic or can cause non-determinism in the state machine.
 //
-// The query can be multi-thread, so we have to use
-// thread safe sync.Map.
-var stargateWhitelist sync.Map
+// The query is multi-threaded so we're using a sync.Pool
+// to manage the allocation and de-allocation of newly created
+// pb objects.
+var stargateResponsePools = make(map[string]*sync.Pool)
```

Comment:
Trying to understand - is the main reason `sync.Pool` works and `sync.Map` doesn't is that the former allocates new objects for concurrent requests?

Reply:
basically the sync.Map was keeping safe the map, which was not needed since after init the map is readonly.

the value of the map was a **pointer** to a **protobuf object**.

So we had a map of `Map[K, *V]`

and simulate + delivertx where all editing the same pointer, meaning they were editing the same variable underneath, concurrently (not the map but the value associated with the key in that map). 

So what were we using that value for? To unmarshal a stargate query into a protobuf object that then we marshal back as JSON for CosmWasm contracts. 

So what this map of sync pool does is that it provides a way to create new objects matching to a specific gRPC query response type (creation=allocation), and when we're done with them we put them it the pool so we can use them again without allocating anymore. sync.Pool takes care of de-allocating them when they're not needed anymore so we do not have to worry.

hope this clarifies the issue

Reply:
My understanding is yes, the later requires a pointer and shares the same struct to unmarshal into, whereas this creates a new object for each request. Utilizing sync.Pool allows us to reallocate the object once completed though so it doesn't have a large impact on performance.

Reply:
testinginprod's explanation is much more thorough, thanks.